### PR TITLE
[JENKINS-67380] Detect oversized maps, collections and arrays and truncate

### PR DESCRIPTION
**Pull request by Vlatombe**
[JENKINS-67380](https://issues.jenkins.io/browse/JENKINS-67380)

- [X] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [X] Ensure that the pull request title represents the desired changelog entry
- [X] Please describe what you did
- [X] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

File: src/main/java/org/jenkinsci/plugins/workflow/cps/actions/ArgumentsActionImpl.java [hunk 1]
```diff
@@ -83,7 +87,7 @@ public ArgumentsActionImpl(@Nonnull Map<String, Object> stepArguments) {
 
     /** For testing use only */
     ArgumentsActionImpl(@Nonnull Set<String> sensitiveVariables){
-        this.isUnmodifiedBySanitization = false;
+        this.isUnmodifiedBySanitization = true;
```
Comment (Vlatombe) [inline src/main/java/org/jenkinsci/plugins/workflow/cps/actions/ArgumentsActionImpl.java:90]: This was wrong. If you do nothing then the data is unmodified.

**Comment by Vlatombe** [comment]
cc @jglick

**Review by jglick — COMMENTED** [review]
Looks like it would fix the issue though I am not sure the details are right. Probably does not matter much either way.

Would prefer to fix https://github.com/jenkinsci/pipeline-utility-steps-plugin/blob/1d91c29606ce59515614ed0d7536b1aaaea9e356/src/main/java/org/jenkinsci/plugins/pipeline/utility/steps/json/WriteJSONStep.java#L55 to be a `String`, perhaps, though this might require some tricks with two JavaBeans attributes backed by a single field. Hard to know exactly what to do with this step because it violates the basic constraint of Pipeline steps that their arguments should be types accepted by `DescribableModel`: primitives (`String`s, true primitives, maybe `enum`s), or `Describable`s and `List<Describable>`s with their attribute types recursively being acceptable; a `Map` is never legal (the `stepArguments` are a special case). Would prefer to just deprecate this step and others like it that deal with nonstandard types; they just cause all sorts of problems. (https://github.com/jenkinsci/pipeline-utility-steps-plugin/pull/47, https://issues.jenkins.io/browse/JENKINS-49669?focusedCommentId=353921&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-353921, etc.)

The fundamental problem with the design of `ArgumentsAction` (as I failed to convince the author of the API at the time) is that it stores the _surface form_ of arguments, raw from Groovy https://github.com/jenkinsci/workflow-cps-plugin/blob/fc3007f2488397da3632908fcd7a2b1b11f79e7f/src/main/java/org/jenkinsci/plugins/workflow/cps/DSL.java#L539 rather than the _logical form_, and you need to call `getResolvedArguments` later and hope you can still reconstruct what was actually meant at the time the build ran. Keeping only the logical form would probably have avoided this particular issue just to the extent that there _is_ no logical form for `WriteJSONStep.json` and so this map entry would simply have been discarded for purposes of `ArgumentsAction`.

File: src/main/java/org/jenkinsci/plugins/workflow/cps/actions/ArgumentsActionImpl.java [hunk 2]
```diff
@@ -295,13 +304,19 @@ Object sanitizeObjectAndRecordMutation(@CheckForNull Object o, @CheckForNull Env
      * Goes through {@link #sanitizeObjectAndRecordMutation(Object, EnvVars)} for each value in a map input.
      */
     @Nonnull
-    Map<String,Object> sanitizeMapAndRecordMutation(@Nonnull Map<String, Object> mapContents, @CheckForNull EnvVars variables) {
+    Object sanitizeMapAndRecordMutation(@Nonnull Map<String, Object> mapContents, @CheckForNull EnvVars variables) {
         // Package scoped so we can test it directly
         LinkedHashMap<String, Object> output = new LinkedHashMap<>(mapContents.size());
+        long size = mapContents.size();
```
[the remark this answers is not in the document]
Comment (Vlatombe) [reply src/main/java/org/jenkinsci/plugins/workflow/cps/actions/ArgumentsActionImpl.java:310]: It's just an arbitrary size measure (~ number of pointers), similar to https://github.com/jenkinsci/workflow-api-plugin/blob/4aa2fc5940f9828c00a5ec71e19250408e644db2/src/main/java/org/jenkinsci/plugins/workflow/actions/ArgumentsAction.java#L97. What actually matters here is we add the actual size of key + value so this gives us a more reliable estimate of the retained size.

[the remark this answers is not in the document]
Comment (jglick) [reply src/main/java/org/jenkinsci/plugins/workflow/cps/actions/ArgumentsActionImpl.java:310]: > actual size of key + value

Just value IIUC? https://github.com/jenkinsci/workflow-cps-plugin/blob/3f47d9098b7a2f027876c39232768a499d4ca6a0/src/main/java/org/jenkinsci/plugins/workflow/cps/actions/ArgumentsActionImpl.java#L313-L315

[the remark this answers is not in the document]
Comment (Vlatombe) [reply src/main/java/org/jenkinsci/plugins/workflow/cps/actions/ArgumentsActionImpl.java:310]: Right, I went through several versions and I missed key in this one. Fixed.

File: src/test/java/org/jenkinsci/plugins/workflow/cps/actions/ArgumentsActionImplTest.java [hunk 3]
```diff
@@ -210,28 +204,74 @@ public void testRecursiveSanitizationOfContent() {
...
+        Assert.assertEquals(sanitizedSmallMap, smallMap);
+        Assert.assertTrue(impl.isUnmodifiedArguments());
+        impl.isUnmodifiedBySanitization = true;
+
+        Map<String, Object> bigMap = new HashMap<>();
+        String bigString = generateStringOfSize(ArgumentsActionImpl.getMaxRetainedLength());
+        bigMap.put("key1", bigString);
+        Object sanitizedBigMap = impl.sanitizeMapAndRecordMutation(bigMap, null);
+        Assert.assertNotEquals(sanitizedBigMap, bigMap);
+        Assert.assertEquals(ArgumentsAction.NotStoredReason.OVERSIZE_VALUE, sanitizedBigMap);
```
[the remark this answers is not in the document]
Comment (Vlatombe) [reply src/test/java/org/jenkinsci/plugins/workflow/cps/actions/ArgumentsActionImplTest.java]: Should be matching what you expect now.

File: src/main/java/org/jenkinsci/plugins/workflow/cps/actions/ArgumentsActionImpl.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (jglick) [reply src/main/java/org/jenkinsci/plugins/workflow/cps/actions/ArgumentsActionImpl.java:320]: Still not counting size of keys? Fine so far as I am concerned, just not sure what you intended.

File: src/main/java/org/jenkinsci/plugins/workflow/cps/actions/ArgumentsActionImpl.java [hunk 4]
```diff
@@ -73,7 +74,10 @@
 
     public ArgumentsActionImpl(@Nonnull Map<String, Object> stepArguments, @CheckForNull EnvVars env, @Nonnull Set<String> sensitiveVariables) {
         this.sensitiveVariables = new HashSet<>(sensitiveVariables);
-        arguments = serializationCheck(sanitizeMapAndRecordMutation(stepArguments, env));
+        Object o = sanitizeMapAndRecordMutation(stepArguments, env);
+        if (o instanceof Map) {
+            arguments = serializationCheck((Map<String, Object>) o);
+        }
```
[the remark this answers is not in the document]
Comment (jglick) [reply src/main/java/org/jenkinsci/plugins/workflow/cps/actions/ArgumentsActionImpl.java]: So this `else` branch should now be an assertion error IIUC?

[the remark this answers is not in the document]
Comment (jglick) [reply src/main/java/org/jenkinsci/plugins/workflow/cps/actions/ArgumentsActionImpl.java]: Sorry, was not looking at latest version of diff.

**Review by jglick — APPROVED** [review]
Looks right. Thanks!

**Review by car-roll — APPROVED** [review]

**Pull request merged**